tests/unit/snap/test_build.py

```diff
@@ -0,0 +1,414 @@
+]
+
+
+@given(strategies.integers(min_value=16, max_value=98).filter(lambda x: x % 2 == 0))
+def test_core_base_regex_match(version):
+    assert _build.CORE_BASE_REGEX.match(f"core{version}")
+
+
+@given(
+    strategies.one_of(
+        strategies.integers(min_value=99),
+        strategies.integers(min_value=16, max_value=98).filter(lambda x: x % 2 != 0),
+        strategies.integers(max_value=15),
+    ),
+)
+def test_core_base_regex_version_non_match(version):
+    assert not _build.CORE_BASE_REGEX.match(f"core{version}")
+
+
+@given(strategies.text().filter(lambda s: not s.startswith("core")))
+def test_core_base_regex_non_match(string):
+    assert not _build.CORE_BASE_REGEX.match(string)
+
+
+@pytest.mark.parametrize(
+    ("string", "matches"),
+    [
+        ("core", True),
+        ("core14", False),
+    ],
+)
+def test_core_base_regex_match_specials(string, matches):
+    assert bool(_build.CORE_BASE_REGEX.match(string)) == matches
+
+
+@pytest.mark.parametrize(
+    ("base", "build_base", "expected"),
+    [
+        ("core", None, craft_platforms.DistroBase("ubuntu", "16.04")),
+        ("core16", None, craft_platforms.DistroBase("ubuntu", "16.04")),
+        ("core18", None, craft_platforms.DistroBase("ubuntu", "18.04")),
+        ("core20", None, craft_platforms.DistroBase("ubuntu", "20.04")),
+        ("core22", None, craft_platforms.DistroBase("ubuntu", "22.04")),
+        ("core22-desktop", "core22", craft_platforms.DistroBase("ubuntu", "22.04")),
+        ("core24", None, craft_platforms.DistroBase("ubuntu", "24.04")),
+        ("core24-desktop", "core24", craft_platforms.DistroBase("ubuntu", "24.04")),
+        ("core26", None, craft_platforms.DistroBase("ubuntu", "26.04")),
+        *(
+            ("bare", f"core{n}", craft_platforms.DistroBase("ubuntu", f"{n}.04"))
+            for n in (18, 20, 22, 24)
+        ),
+    ],
+)
+def test_get_distro_base_from_core_base_success(base, build_base, expected):
+    assert _build.get_distro_base_from_core_base(base, build_base) == expected
+
+
+@pytest.mark.parametrize(
+    ("base", "build_base", "exc_class", "exc_match"),
+    [
+        (
+            "bare",
+            "devel",
+            craft_platforms.InvalidBaseError,
+            "build-base 'devel' is not valid if base is 'bare'",
+        ),
+        (
+            "core22-desktop",
+            None,
+            craft_platforms.NeedBuildBaseError,
+            "base 'core22-desktop' requires a 'build-base'",
+        ),
+        (
+            "bare",
+            None,
+            craft_platforms.NeedBuildBaseError,
+            "base 'bare' requires a 'build-base'",
+        ),
+        (
+            "core22",
+            "core22",
+            craft_platforms.InvalidBaseError,
+            "base 'core22' cannot use a build-base",
+        ),
+        (
+            "devel",
+            None,
+            craft_platforms.InvalidBaseError,
+            "base 'devel' is unknown or invalid",
+        ),
```

Comment:
This build planner is awesome, nice work!  Much cleaner and more correct than what's in snapcraft today.